# Test Stage B: Machine Learning: Regression - Predicting Energy Efficiency of Buildings

In [27]:
#Importing packages
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso

In [3]:
#Importing dataset
df = pd.read_csv('energydata_complete.csv')
df.head(10)

date  Appliances  lights         T1       RH_1     T2  \
0  2016-01-11 17:00:00          60      30  19.890000  47.596667  19.20   
1  2016-01-11 17:10:00          60      30  19.890000  46.693333  19.20   
2  2016-01-11 17:20:00          50      30  19.890000  46.300000  19.20   
3  2016-01-11 17:30:00          50      40  19.890000  46.066667  19.20   
4  2016-01-11 17:40:00          60      40  19.890000  46.333333  19.20   
5  2016-01-11 17:50:00          50      40  19.890000  46.026667  19.20   
6  2016-01-11 18:00:00          60      50  19.890000  45.766667  19.20   
7  2016-01-11 18:10:00          60      50  19.856667  45.560000  19.20   
8  2016-01-11 18:20:00          60      40  19.790000  45.597500  19.20   
9  2016-01-11 18:30:00          70      40  19.856667  46.090000  19.23   

        RH_2     T3       RH_3         T4  ...         T9   RH_9     T_out  \
0  44.790000  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000   
1  44.722500  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333   
2  44.626667  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667   
3  44.590000  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000   
4  44.530000  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333   
5  44.500000  19.79  44.933333  18.890000  ...  17.000000  45.29  6.016667   
6  44.500000  19.79  44.900000  18.890000  ...  17.000000  45.29  5.900000   
7  44.500000  19.73  44.900000  18.890000  ...  17.000000  45.29  5.916667   
8  44.433333  19.73  44.790000  18.890000  ...  17.000000  45.29  5.933333   
9  44.400000  19.79  44.863333  18.890000  ...  17.000000  45.29  5.950000   

   Press_mm_hg     RH_out  Windspeed  Visibility  Tdewpoint        rv1  \
0   733.500000  92.000000   7.000000   63.000000   5.300000  13.275433   
1   733.600000  92.000000   6.666667   59.166667   5.200000  18.606195   
2   733.700000  92.000000   6.333333   55.333333   5.100000  28.642668   
3   733.800000  92.000000   6.000000   51.500000   5.000000  45.410389   
4   733.900000  92.000000   5.666667   47.666667   4.900000  10.084097   
5   734.000000  92.000000   5.333333   43.833333   4.800000  44.919484   
6   734.100000  92.000000   5.000000   40.000000   4.700000  47.233763   
7   734.166667  91.833333   5.166667   40.000000   4.683333  33.039890   
8   734.233333  91.666667   5.333333   40.000000   4.666667  31.455702   
9   734.300000  91.500000   5.500000   40.000000   4.650000   3.089314   

         rv2  
0  13.275433  
1  18.606195  
2  28.642668  
3  45.410389  
4  10.084097  
5  44.919484  
6  47.233763  
7  33.039890  
8  31.455702  
9   3.089314  

[10 rows x 29 columns]

## Question 12

In [6]:
#linear regression model
model = LinearRegression()

#T2 = Celsius and T6 = temperature outside the building 
X, y = df['T6'], df['T2']
X = X[:, None]

#fit regression model
model.fit(X, y)

#calculate R-squared of regression model
r_squared = model.score(X, y)

#view R-squared value
print(r_squared)


0.6418990830855493


<ipython-input-6-836690ca6e9e>:8: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  X = X[:, None]


The coefficient of determination is 0.642. The 64% of the response variable  is explained by the model.

## Question 13

In [15]:
# Prepare the normalization tool
scaler = MinMaxScaler()

#drop columns
df_n=df.drop(columns=[ 'date', 'lights'])

# Normalize the data 
normalised_df = pd.DataFrame(scaler.fit_transform(df_n), columns=df_n.columns)
features_df=normalised_df.drop(columns=[ 'Appliances'])
target = normalised_df[ 'Appliances' ]

#train-test set split

x_train, x_test, y_train, y_test = train_test_split(features_df, target,
test_size= 0.3 , random_state= 42 )

#set the model
linear_model = LinearRegression()

#fit the model to the training dataset
linear_model.fit(x_train, y_train)

#predictions
predicted_y = linear_model.predict(x_test)

#MAE
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(y_test, predicted_y)
round(mae, 2 ) 

0.05

## Question 14

In [14]:
#Residual Sum of Squares 
rss = np.sum(np.square(y_test - predicted_y))
round(rss, 2 ) 

45.35

## Question 15

In [17]:
# Root Mean Squared Error 

rmse = np.sqrt(mean_squared_error(y_test, predicted_y))
round(rmse, 3) 

0.088

## Question 16

In [22]:
#Coeficient of Determination 

r2_score = r2_score(y_test, predicted_y)
round(r2_score, 2) 

0.15

## Question 17

In [46]:
linear_weights = pd.Series(linear_model.coef_, x_train.columns).sort_values()
linear_weightsdf = pd.DataFrame(linear_weights).reset_index()
linear_weightsdf.columns = [ 'Features' , 'Linear_Model']
linear_weightsdf['Linear_Model']=linear_weightsdf['Linear_Model'].round(3)
linear_weightsdf

Features  Linear_Model
0          RH_2        -0.457
1         T_out        -0.322
2            T2        -0.236
3            T9        -0.190
4          RH_8        -0.158
5        RH_out        -0.078
6          RH_7        -0.045
7          RH_9        -0.040
8            T5        -0.016
9            T1        -0.003
10          rv1         0.001
11          rv2         0.001
12  Press_mm_hg         0.007
13           T7         0.010
14   Visibility         0.012
15         RH_5         0.016
16         RH_4         0.026
17           T4         0.029
18    Windspeed         0.029
19         RH_6         0.038
20         RH_3         0.096
21           T8         0.102
22    Tdewpoint         0.118
23           T6         0.236
24           T3         0.291
25         RH_1         0.554

The lowest: RH_2, T_out , the highest: RH_1, T3. The highest: TH1 and the lowest: RH_2.

## Question 18

In [26]:
#ridge
ridge_reg = Ridge(alpha= 0.4 )
ridge_reg.fit(x_train, y_train)

predicted_y_r = linear_model.predict(x_test)

# root mean squared error (RMSE) 
rmse = np.sqrt(mean_squared_error(y_test, predicted_y_r))
round(rmse, 3) 

0.088

## Question 19

In [54]:
#lasso
lasso_reg = Lasso(alpha= 0.001 )
lasso_reg.fit(x_train, y_train)

#lasso weight
lasso_weights = pd.Series(lasso_reg.coef_, x_train.columns).sort_values()
lasso_weightsdf = pd.DataFrame(lasso_weights).reset_index()
lasso_weightsdf.columns = [ 'Features' , 'lasso']
lasso_weightsdf['lasso']=lasso_weightsdf['lasso'].round(4)

#count
c=0
for i in lasso_weightsdf['lasso']:
    if i != float(0.0000):
        c +=1
    else:
        c +=0
print(c)

4


## Question 20

In [55]:
predicted_y_l = lasso_reg.predict(x_test)

# root mean squared error (RMSE) 
rmse = np.sqrt(mean_squared_error(y_test, predicted_y_l))
round(rmse, 3) 

0.094